spark_log_parser/loaders/__init__.py

```diff
@@ -0,0 +1,282 @@
+import abc
+import gzip
+import tarfile
+import logging
+
+from io import IOBase, BufferedIOBase
+from pathlib import Path
+from urllib.parse import ParseResult
+
+from aiodataloader import DataLoader
+from stream_unzip import stream_unzip
+
+FILE_SKIP_PATTERNS = [".DS_Store".lower(), "__MACOSX".lower(), "/."]
+
+
+class FileChunkStreamWrapper:
+    """
+    Minimal file-like wrapper to wrap the output of file streams that may be lacking methods like read/readlines (or in
+    other words, file streams that are raw generators). This allows us to wrap things like the output of `stream_unzip`
+    or botocore.StreamingBody.iter_lines() and pass those streams around as file-like objects.
+
+    Note - this wrapper does not implement the full file-like object API. Instead, it only implements at the moment
+        those methods that are actually called when loading files in various contexts. This doesn't mean that
+        new methods won't need to be implemented in the future as use-cases grow, it just means that at the moment
+        this minimal implementation satisfies the current use-cases we do have
+
+    Note #2 - once `chunks` is consumed, calls to read/lines will return empty byte buffers
+    """
+
+    _trailing_data: bytes = b''
+    _chunks = None
+
+    def __init__(self, chunks):
+        assert chunks
+        self._chunks = chunks
+
+    def __iter__(self):
+        """
+        The implementation of
+        """
+        return self.readlines()
+
+    def read(self, size=-1):
+        content = self._trailing_data
+        self._trailing_data = b''
+
+        if not size or size == -1:
+            for chunk in self._chunks:
+                content += chunk
+        else:
+            # iterate chunks until we've read at least `size` amount of data
+            for chunk in self._chunks:
+                content += chunk
+                if len(content) >= size:
+                    break
+
+            # Trim off any data that is past the provided `size` and hold on to it for subsequent `read` calls
+            if len(content) > size:
+                self._trailing_data = content[size:]
+                content = content[:size]
+
+        return content
+
+    def readlines(self):
+        """
+        Returns a stream of lines from self.chunks. This differs from many readlines() implementations, which tend to
+        read the whole file into memory and return a list of the lines
+        """
+        pending = None
+
+        for chunk in self._chunks:
+
+            if pending is not None:
+                chunk = pending + chunk
+
+            lines = chunk.splitlines()
+
+            # Check if the last item in our last line is the same as the last item in our original un-split line.
+            #  If it is, that means our original chunk did not end in a newline and we need to hang on to that data
+            #  for processing the next chunk
+            if lines and lines[-1] and chunk and lines[-1][-1] == chunk[-1]:
```

Comment:
This seems more complicated than it has to be, and if chunk size happened to align with line length you'd end up using more memory than expected. Instead, how about just,
```
if chunk[-1] != "\n":
    pending = lines.pop()
```

Reply:
yeah, I actually scooped this implementation up from how `requests` implements `iter_lines()` and handles trailing data. This is a good point, though. I'm trying to think of why just `if chunk and chunk[-1] != "\n"` wouldn't work... it's not obvious to me why it wouldn't, unless of course `splitlines()` is splitting on a character sequence other than `"\n"`

Reply:
While attempting to add back the tar/zip-bomb protecting, I actually ran into a case where `chunk[-1] != "\n"` doesn't appear to work - if the underlying `chunk` is some `bytes` (vs. a String), then `chunk[-1]` will equal the Integer 10, but not be equal to "\n" - 

<img width="288" alt="Screenshot 2022-12-19 at 4 15 35 PM" src="https://user-images.githubusercontent.com/117235708/208552372-a24f48b6-22b7-470d-b92a-c4fc6a27499f.png">

The original check handles this appropriately, though, and so seems to be agnostic of the underlying data type (whether that's bytes, or an actual String, or something else entirely)

Reply:
Ah. `readlines` suggests a character stream to me (lines of text?) which might allow us to make that simplification... Assuming we don't want to do any decoding here I feel good borrowing from `requests`'s `iter_lines` - tried n' true. A type descriptor for `_chunks` or an additional comment on the statement that conditionally pops the last line might help if others find their way to this code.